### PR TITLE
[IMPROVE] SAML login process refactoring

### PR DESCRIPTION
**Pull request by kukkjanos**
- the eppn is primary identifier match parameter (if not exist then email is a secondary)
- update eppn to user profile
- support the eduPersonPrincipalName (eppn) usually required saml parameter
- support the displayName optional saml parameter
- support two new saml settings
  - overwrite user fullname if need (use idp attribute [cn,username or displayName])
  - overwrite user mail address if need (use idp attribute)

FIX NEW DIRECTORY

Closes #ISSUE_NUMBER

### CODE REVIEW

**Comment by northway** [comment]
**Could someone approve this pull request?**

I applied this code by hand and it works!

**Comment by kukkjanos** [comment]
Hi @rodrigok!

Please help me to move on to the next state.
Is there a known problem?

Thanks,

File: app/meteor-accounts-saml/server/saml_server.js [hunk 1]
```diff
@@ -111,14 +113,39 @@ Accounts.registerLoginHandler(function(loginRequest) {
 	if (loginResult && loginResult.profile && loginResult.profile.email) {
 		const emailList = Array.isArray(loginResult.profile.email) ? loginResult.profile.email : [loginResult.profile.email];
 		const emailRegex = new RegExp(emailList.map((email) => `^${ RegExp.escape(email) }$`).join('|'), 'i');
+
+		const eduPersonPrincipalName = loginResult.profile.eppn;
+		const fullName = loginResult.profile.cn || loginResult.profile.username || loginResult.profile.displayName;
+
+		let eppnMatch = false;
+		// not used
+		// let emailMatch = false;
```
Comment (Hudell) [inline app/meteor-accounts-saml/server/saml_server.js]: Please remove the commented code.

File: app/meteor-accounts-saml/server/saml_server.js [hunk 2]
```diff
@@ -111,14 +113,39 @@ Accounts.registerLoginHandler(function(loginRequest) {
...
+
+		// Check eppn
 		let user = Meteor.users.findOne({
-			'emails.address': emailRegex,
+			eppn: eduPersonPrincipalName,
 		});
 
+		if (user) {
+			console.log('eppn match');
```
Comment (Hudell) [inline app/meteor-accounts-saml/server/saml_server.js]: Please remove the console.log call.

File: app/meteor-accounts-saml/server/saml_server.js [hunk 3]
```diff
@@ -111,14 +113,39 @@ Accounts.registerLoginHandler(function(loginRequest) {
...
+			eppnMatch = true;
+		}
+
+		// If eppn is not exist
+		if (!user) {
+			user = Meteor.users.findOne({
+				'emails.address': emailRegex,
+			});
+
+			console.log('email match');
```
Comment (Hudell) [inline app/meteor-accounts-saml/server/saml_server.js]: Please remove the console log call.

File: app/meteor-accounts-saml/server/saml_server.js (hunk 2 above)
Comment (kukkjanos) [reply app/meteor-accounts-saml/server/saml_server.js]: Fixed

File: app/meteor-accounts-saml/server/saml_server.js (hunk 3 above)
Comment (kukkjanos) [reply app/meteor-accounts-saml/server/saml_server.js]: Fixed

File: app/meteor-accounts-saml/server/saml_server.js (hunk 1 above)
Comment (kukkjanos) [reply app/meteor-accounts-saml/server/saml_server.js]: Fixed

**Comment by Toushe** [comment]
+1

**Comment by Hudell** [comment]
Does it need a different authnContext value to work? What value would that be?

**Comment by rAcHekLoS** [comment]
request += '<samlp:RequestedAuthnContext xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Comparison="exact">' + '<saml:AuthnContextClassRef xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">urn:federation:authentication:windows</saml:AuthnContextClassRef></samlp:RequestedAuthnContext>\n' + '</samlp:AuthnRequest>';

**Comment by isometry** [comment]
> a) This change destroy SAML with Microsoft ADFS. The login works, but the user maps always to the User Rocket.Cat.

This change also breaks SAML with OneLogin IdP; same symptoms as reported here, but OneLogin *does* support the hard-coded authnContext of `urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport`

**Comment by Hudell** [comment]
You mean login working but with the wrong user? That was fixed by #14686 and included on the 1.1.1 hotfix release.

**Comment by isometry** [comment]
> You mean login working but with the wrong user? That was fixed by #14686 and included on the 1.1.1 hotfix release.

Confirmed fixed. Thanks!

**Comment by rAcHekLoS** [comment]
The problem #14686 is fixed, THX.

For AuthContext, I would suggest that you make an array to specify multiple AuthContext. This way you can support multiple variants.

**Pull request merged**